contrib/platform/src/com/sun/jna/platform/win32/WbemcliUtil.java

```diff
@@ -0,0 +1,783 @@
+     *            properties
+     * @return a WmiResult object containing the query results, wrapping an
+     *         EnumMap
+     */
+    public static <T extends Enum<T>> WmiResult<T> queryWMI(WmiQuery<T> query) {
+        try {
+            return queryWMI(query, Wbemcli.WBEM_INFINITE);
+        } catch (TimeoutException e) {
+            throw new WbemcliException("Got a WMI timeout when infinite wait was specified. This should never happen.",
+                    Wbemcli.WBEM_INFINITE);
+        }
+    }
+
+    /**
+     * Query WMI for values, with a specified timeout.
+     * 
+     * @param <T>
+     *            an enum
+     * @param query
+     *            A WmiQuery object encapsulating the namespace, class, and
+     *            properties
+     * @param timeout
+     *            Number of milliseconds to wait for results before timing out.
+     *            If {@link IEnumWbemClassObject#WBEM_INFINITE} (-1), will
+     *            always wait for results. If a timeout occurs, throws a
+     *            {@link TimeoutException}.
+     * @return a WmiResult object containing the query results, wrapping an
+     *         EnumMap
+     * @throws TimeoutException
+     *             if the query times out before completion
+     */
+    public static <T extends Enum<T>> WmiResult<T> queryWMI(WmiQuery<T> query, int timeout) throws TimeoutException {
+        // Idiot check
+        if (query.getPropertyEnum().getEnumConstants().length < 1) {
+            throw new WbemcliException("The query's property enum has no values.",
+                    query.getPropertyEnum().getEnumConstants().length);
+        }
+
+        // Initialize COM if not already done. Needed if COM was previously
+        // initialized externally but is no longer initialized.
+        if (!isComInitialized()) {
+            initCOM();
+        }
+
+        // Connect to the server
+        IWbemServices svc = connectServer(query.getNameSpace());
+
+        // Send query
+        IEnumWbemClassObject enumerator = selectProperties(svc, query);
+
+        try {
+            return enumerateProperties(enumerator, query.getPropertyEnum(), timeout);
+        } catch (TimeoutException e) {
+            throw new TimeoutException(e.getMessage());
+        } finally {
+            // Cleanup
+            enumerator.Release();
+            svc.Release();
+        }
+    }
+
+    /*
+     * Below methods ported from: Getting WMI Data from Local Computer
+     * https://docs.microsoft.com/en-us/windows/desktop/WmiSdk/example--getting-
+     * wmi-data-from-the-local-computer
+     *
+     * Steps 1 - 7 in the comments correspond to the above link. Steps 1 through
+     * 5 contain all the steps required to set up and connect to WMI, and steps
+     * 6 and 7 are where data is queried and received.
+     */
+
+    /**
+     * Initializes COM library and sets security to impersonate the local user
+     */
+    public static void initCOM() {
```

Comment:
This looks dangerous if you combine it with other COM calling code.

Reply:
I agree it looks dangerous but I'm not sure how else to handle it.  I can ignore "already initialized" errors but then if I uninit later, it will kill the external program's COM code.  If I don't uninit, it's a resource leak.

Reply:
Make the user responsible for initializing COM. Else you need to guess how the user will use it. The COM Utilities have offer both options. COMThread handles the creation of a thread for dispatching COM Threads with correctly initialized COM. `Factory` uses that, while `ObjectFactory` allows creation of call proxies without a different thread.

I found it better to handle COM initialization myself, so I prefer to use the `ObjectFactory`.

There is a problem with your code - when calls to the utility methods are done from different threads, COM will only be initialized once. This violates the requirement, that `CoInitializeEx` must be called on each thread making COM calls.

Reply:
I've moved the COM outside the Util class.